research/05-problem-ana.tex

```diff
@@ -1,10 +1,127 @@
+whether this will be present in the final product.
 
-\subsection{How does REPL generation fit within Spoofax}
-\label{ssec:architecture}
+\subsection{Incorporating REPL generation within Spoofax}
+\label{ssec:incorp-repl-gener}
+To generate solidly working REPLs for any language defined in Spoofax,
+some Spoofax concepts have to be carefully considered. Since every
+language can have many different language constructs, generating REPLs
+should not be constrained to language specific constructs like
+classes, but should rather be done generically. Therefore the
+implementation and particularly the interaction with the user needs to
+be thought of carefully.
 
-\subsection{How are plugins developed for Eclipse?}
+This section lists the problems that are expected when trying to
+implement generating REPLs generically. For each problem, their
+solution approaches that should be explored during the implementation
+process are given, as well as the parts of Spoofax that are relevant
+for that problem.
+
+\subsubsection{Detecting unfinished expressions for multiline editing}
+\label{sec:detect-unfin-expr}
+To support multiline editing, the generated REPL should detect that
+the expression is unfinished when the user presses ``enter'' for the
+next line, instead of trying to parse and execute the unfinished
+expression. An obvious part of Spoofax that is relevant for this
+problem, is the syntax definition in SDF3 (\cref{sec:orgheadline1}).
+
+\subsubsection{Language specific additional commands}
+\label{sec:lang-spec-addit}
+Another problem is when some additional command can be useful for a
+REPL of a particular language, but would not make any sense within the
+context of other languages. For example, some languages such as Java
+have the notion of a ``module'' or ``package'', which declares a
+separate namespace for the definitions in it. An additional command to
+list all of the currently loaded modules is useful, but would not make
+any sense in languages that have no concept of a module.
+
+This problem could be solved by the language designer by extending his
+or her language with reflective capabilities. However in that case it
+would not be an additional command, but just an expression like any
+other expression in that language. Moreover, the language designer
+might not want to extend the language with reflective capabilities
+outside of the context of a REPL, for example when the language is a
+DSL. It is clear therefore that a different approach should be
+considered.
+
+\begin{figure}[htbp]
+  \centering
+  \includegraphics[width=0.5\textwidth]{menu-actions.png}
+  \caption{A menu action for the paplj language defined using
+    Spoofax. The bottom window shows the menu definition, the top
+    window shows a program written in paplj.}
+  \label{fig:menu-actions}
+\end{figure}
+
+One possible solution is to allow for language specific configurations
+that are loaded and used during the generation process of a REPL for
```

Comment:
No I don't think so, because it is a specific feature of the REPL right? Don't think it really matters though, it's more about the ability to configure.